### PR TITLE
add new test messages for arrays of arrays

### PR DESCRIPTION
**Pull request by Karsten1987**
We'd like to have two more test messages in the form of containers of containers. We would leverage them in our rosbag tests.

see https://github.com/ros2/rosbag2/pull/56#issuecomment-439093754
cc @anhosi

### CODE REVIEW

File: test_msgs/msg/DynamicArrayDynamicArrayPrimitivesNested.msg [hunk 1]
```diff
@@ -0,0 +1 @@
+DynamicArrayPrimitives[] dynamic_array_dynamic_array_primitive_values
```
Comment (jacobperron) [inline test_msgs/msg/DynamicArrayDynamicArrayPrimitivesNested.msg]: This case already exists, equivalent to `DynamicArrayPrimitivesNested`:

https://github.com/ros2/rcl_interfaces/blob/2a85883cf64d9dfbe9f15a43966bdb00819aac58/test_msgs/msg/DynamicArrayPrimitivesNested.msg#L1

File: test_msgs/CMakeLists.txt [hunk 2]
```diff
@@ -21,9 +21,11 @@ find_package(action_msgs REQUIRED)
 find_package(rosidl_default_generators REQUIRED)
 
 set(msg_files
+  "msg/BoundedArrayBoundedArrayPrimitivesNested.msg"
```
Comment (jacobperron) [inline test_msgs/CMakeLists.txt]: For consistency, I would name this `BoundedArrayPrimitivesNested`.

**Review by jacobperron — CHANGES_REQUESTED** [review]
There are some consequences to adding new messages to this package:

1. It will cause [test_communications](https://github.com/ros2/system_tests/tree/master/test_communication) to fail, unless a few things are updated.
See my tickets related to previous tribulations:

- #45 
- ros2/system_tests#302

2. The build time increases significantly for all builds (exponential?) because of testing done in `test_communications`.
(@clalancette Correct me if I'm wrong).

**Comment by Karsten1987** [comment]
Comments make sense to me, but I'll let @anhosi respond. 

If this doesn't get merged because of runtime reasons, we might have our own `rosbag2_test_msgs` paket. But i'd prefer not to.

**Comment by jacobperron** [comment]
IMO, it makes sense to add the missing cases to this package for completeness:

- BoundedArrayPrimitivesNested
- StaticArrayPrimitivesNested

File: test_msgs/msg/DynamicArrayDynamicArrayPrimitivesNested.msg (hunk 1 above)
Comment (anhosi) [reply test_msgs/msg/DynamicArrayDynamicArrayPrimitivesNested.msg]: You are right, we missed this one.

File: test_msgs/CMakeLists.txt (hunk 2 above)
Comment (anhosi) [reply test_msgs/CMakeLists.txt]: Makes sense, our suggestion was motivated by `DynamicArrayStaticArrayPrimitivesNested.msg`. But I would prefer your shorter suggestion.

**Comment by Karsten1987** [comment]
I just updated the message files. Is there anything else to do with it in terms of test_communication?

**Comment by jacobperron** [comment]
> I just updated the message files. Is there anything else to do with it in terms of test_communication?

@Karsten1987 
Support should be added for the new message types to the utility functions in this package:
- [message_fixtures.py](https://github.com/ros2/rcl_interfaces/blob/c48d2731548b36b1e8458e37f4f84d187ed8ed65/test_msgs/src/test_msgs/message_fixtures.py#L323)
- [message_fixtures.hpp](https://github.com/ros2/rcl_interfaces/blob/master/test_msgs/include/test_msgs/message_fixtures.hpp)

And then the publisher/subscriber tests in test_communications should be updated:

- https://github.com/ros2/system_tests/blob/5267cdd2ff2bb40adcee02e28f2c13901e51b6a2/test_communication/test/test_subscriber.cpp#L105
- https://github.com/ros2/system_tests/blob/5267cdd2ff2bb40adcee02e28f2c13901e51b6a2/test_communication/test/test_publisher_subscriber.cpp#L135
- https://github.com/ros2/system_tests/blob/5267cdd2ff2bb40adcee02e28f2c13901e51b6a2/test_communication/test/test_publisher.cpp#L73

An alternative to all of these changes is to let the publisher/subscriber tests in test_communications pass if the message type is unknown. Then any future additions to `test_msgs` wouldn't break integration tests. I.e. Return 0 [here](https://github.com/ros2/system_tests/blob/5267cdd2ff2bb40adcee02e28f2c13901e51b6a2/test_communication/test/test_subscriber.cpp#L144) and the same for the other test I've linked.

**Comment by Karsten1987** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5586/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2220/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4595/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5465/)

**Review by jacobperron — APPROVED** [review]

**Comment by Karsten1987** [comment]
new round of CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/5589/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2223/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/4598/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/5468/)

**Pull request merged**